### PR TITLE
Allow opting out of model nesting by setting `model=None`

### LINKED ISSUE #6715: DOC: Provide a working example of inheriting from pymc.Model
### Issue with current documentation:

The documentation for [`pymc.Model`](https://www.pymc.io/projects/docs/en/stable/api/generated/pymc.Model.html) seems to give some conceptual gist of how it works, but the example given in the documentation [does not work](https://stackoverflow.com/questions/76224256/what-is-the-model-argument-in-the-super-init-in-the-example-subclassing/76224316?noredirect=1#comment134420040_76224316).

There are no imports, and I am not sure where `pt` is coming from in `pt.constant(1)`.  I have thought maybe `pt` was for "PyTensor" but there is no `constant` in the namespace of `pt` resulting from `import pytensor as pt`. There is a `pymc.math.constant`, but I am unsure if that is the same thing.

But it is also unclear if the imports are the only issue. Here is some code where I have attempted to address the above namespace issues.

```python
from pymc import *

class CustomModel(Model):
    # 1) override init
    def __init__(self, mean=0, sigma=1, name=''):
        # 2) call super's init first, passing model and name
        # to it name will be prefix for all variables here if
        # no name specified for model there will be no prefix
        super().__init__(name, model)
        # now you are in the context of instance,
        # `modelcontext` will return self you can define
        # variables in several ways note, that all variables
        # will get model's name prefix

        # 3) you can create variables with the register_rv method
        self.register_rv(Normal.dist(mu=mean, sigma=sigma), 'v1', initval=1)
        # this will create variable named like '{name::}v1'
        # and assign attribute 'v1' to instance created
        # variable can be accessed with self.v1 or self['v1']

        # 4) this syntax will also work as we are in the
        # context of instance itself, names are given as usual
        Normal('v2', mu=mean, sigma=sigma)

        # something more complex is allowed, too
        half_cauchy = HalfCauchy('sigma', beta=10, initval=1.)
        Normal('v3', mu=mean, sigma=half_cauchy)

        # Deterministic variables can be used in usual way
        Deterministic('v3_sq', self.v3 ** 2)

        # Potentials too
        Potential('p1', pt.constant(1))

# After defining a class CustomModel you can use it in several
# ways

# I:
#   state the model within a context
with Model() as model:
    CustomModel()
    # arbitrary actions

# II:
#   use new class as entering point in context
with CustomModel() as model:
    Normal('new_normal_var', mu=1, sigma=0)

# III:
#   just get model instance with all that was defined in it
model = CustomModel()

# IV:
#   use many custom models within one context
with Model() as model:
    CustomModel(mean=1, name='first')
    CustomModel(mean=2, name='second')

# variables inside both scopes will be named like `first::*`, `second::*`
```

Running the above I get an `AttributeError` that I don't think is related because it occurs at `super().__init__(name, model)`.

```python
Traceback (most recent call last):
  File "/usr/lib/python3.10/idlelib/run.py", line 578, in runcode
    exec(code, self.locals)
  File "/home/galen/pymc_inheritance_question.py", line 42, in <module>
    CustomModel()
  File "/home/galen/.local/lib/python3.10/site-packages/pymc/model.py", line 264, in __call__
    instance.__init__(*args, **kwargs)
  File "/home/galen/pymc_inheritance_question.py", line 10, in __init__
    super().__init__(name, model)
  File "/home/galen/.local/lib/python3.10/site-packages/pymc/model.py", line 588, in __init__
    self.add_coords(coords)
  File "/home/galen/.local/lib/python3.10/site-packages/pymc/model.py", line 1080, in add_coords
    for name, values in coords.items():
AttributeError: 'Model' object has no attribute 'items'
```

### Idea or request for content:

Please provide a minimum working example.

### PR DESCRIPTION
**Pull request by ricardoV94**
## Description

This PR allows setting `model=None` to explicitly disable model nesting. This is needed for model transformations `do, observe, clone, remove_value_transforms, freeze_rv_and_dims` to be possible under a model context without side-effects on the original model.

I still think model nesting is an overkill, but a lot of code out there depends on it so I am not touching it.

I also updated the docstrings with more relevant (imo) examples. The previous example was broken as per #6715 and I am not sure we really want to incentivize people to subclass PyMC models. That certainly shouldn't be the focus of the docstrings?

Finally, removed deprecated functionality.

## Related Issue

- [x] Closes #6715 
- [x] Related to #7268 
- [x] Related to https://github.com/pymc-devs/pymc-experimental/pull/345#discussion_r1627311093
- [x] Related to #7348

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [x] Added necessary documentation (docstrings and/or example notebooks)
- [x] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [x] New feature / enhancement
- [ ] Bug fix
- [x] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

----
📚 Documentation preview 📚: https://pymc--7352.org.readthedocs.build/en/7352/

### CODE REVIEW

File: pymc/sampling/deterministic.py [hunk 1]
```diff
@@ -83,7 +83,7 @@ def compute_deterministics(
     model = modelcontext(model)
 
     if var_names is None:
-        deterministics = model.deterministics
+        deterministics = list(model.deterministics)
```
Comment (ricardoV94) [inline pymc/sampling/deterministic.py:86]: To appease mypy

File: pymc/stats/log_density.py [hunk 2]
```diff
@@ -126,46 +128,35 @@ def compute_log_density(
     if kind not in ("likelihood", "prior"):
         raise ValueError("kind must be either 'likelihood' or 'prior'")
 
+    # We need to disable transforms, because the InferenceData only keeps the untransformed values
+    umodel = remove_value_transforms(model)
```
Comment (ricardoV94) [inline pymc/stats/log_density.py:132]: This kind of change is why this PR was done. A similar thing was taking place in https://github.com/pymc-devs/pymc-experimental/pull/345#discussion_r1627311093

File: pymc/model/core.py [hunk 3]
```diff
@@ -400,76 +393,99 @@ class Model(WithMemoization, metaclass=ContextMeta):
     name : str
         name that will be used as prefix for names of all random
         variables defined within model
+    coords: dict
```
Comment (OriolAbril) [inline pymc/model/core.py]:
```suggestion
    coords : dict
```

File: pymc/model/core.py [hunk 4]
```diff
@@ -400,76 +393,99 @@ class Model(WithMemoization, metaclass=ContextMeta):
...
+        is created inside another model's context, the parent model will be set to that model.
+        If `None` the model will not have a parent.
 
     Examples
     --------
-    How to define a custom model
+    Use context manager to define model and respective variables
 
     .. code-block:: python
+        import pymc as pm
```
Comment (OriolAbril) [inline pymc/model/core.py]: I think you need a blank line for this to render correctly. The preview is missing these code blocks: https://pymcio--7352.org.readthedocs.build/projects/docs/en/7352/api/model/generated/pymc.model.core.Model.html

File: pymc/model/core.py [hunk 5]
```diff
@@ -400,76 +393,99 @@ class Model(WithMemoization, metaclass=ContextMeta):
...
+
+        model = pm.Model()
+        x = pm.Normal("x", model=model)
+
+
+    Use coords for defining the shape of random variables and labeling other model variables
+
+    .. code-block:: python
+    import pymc as pm
```
Comment (OriolAbril) [inline pymc/model/core.py]: Here in addition to the black line the code itself must be indented

File: pymc/model/core.py [hunk 6]
```diff
@@ -400,76 +393,99 @@ class Model(WithMemoization, metaclass=ContextMeta):
...
+    import numpy as np
+
+    coords = {"feature", ["A", "B", "C"], "trial", [1, 2, 3, 4, 5]}
```
Comment (michaelosthege) [inline pymc/model/core.py]:
```suggestion
    coords = {
        "feature", ["A", "B", "C"],
        "trial", [1, 2, 3, 4, 5],
    }
```
a little easier to read

**Review by michaelosthege — APPROVED** [review]
I wonder how `.to_graphviz` looks for hierarchical models 🤔

Please apply the docstring fixes that Oriol suggested 👍

**Comment by ricardoV94** [comment]
> I wonder how .to_graphviz looks for hierarchical models 🤔

Doesn't do anything different for the outer model. You can just call `.to_graphviz` on submodels and will only show the variables that belong to it IIRC

**Pull request merged**